File: packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/HandlerBinaryMessenger.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "HandlerBinaryMessenger.h"

@interface HandlerBinaryMessenger ()
@property(nonatomic, strong) NSObject<FlutterMessageCodec> *codec;
@property(nonatomic, copy) HandlerBinaryMessengerHandler handler;
@end

@implementation HandlerBinaryMessenger {
  int _count;
}

- (instancetype)initWithCodec:(NSObject<FlutterMessageCodec> *)codec
                      handler:(HandlerBinaryMessengerHandler)handler {
  self = [super init];
  if (self) {
    _codec = codec;
    _handler = [handler copy];
  }
  return self;
}

- (void)cleanupConnection:(FlutterBinaryMessengerConnection)connection {
}

- (void)sendOnChannel:(nonnull NSString *)channel message:(NSData *_Nullable)message {
}

- (void)sendOnChannel:(nonnull NSString *)channel
              message:(NSData *_Nullable)message
          binaryReply:(FlutterBinaryReply _Nullable)callback {
  NSArray *args = [self.codec decode:message];
  id result = self.handler(args);
  callback([self.codec encode:result]);
}

- (FlutterBinaryMessengerConnection)setMessageHandlerOnChannel:(nonnull NSString *)channel
                                          binaryMessageHandler:
                                              (FlutterBinaryMessageHandler _Nullable)handler {
  return ++_count;
}

@end

File: packages/pigeon/platform_tests/test_plugin/example/ios/RunnerTests/HandlerBinaryMessenger.swift
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation
import Flutter

typealias HandlerBinaryMessengerHandler = ([Any?])-> Any

class HandlerBinaryMessenger: NSObject, FlutterBinaryMessenger {
  let codec: FlutterMessageCodec
  let handler: HandlerBinaryMessengerHandler
  private var count = 0
  
  init(codec: FlutterMessageCodec, handler: @escaping HandlerBinaryMessengerHandler) {
    self.codec = codec
    self.handler = handler
    super.init()
  }
  
  func send(onChannel channel: String, message: Data?) {
    // Method not implemented because this messenger is just for handling
  }
  
  func send(
    onChannel channel: String,
    message: Data?,
    binaryReply callback: FlutterBinaryReply? = nil
  ) {
    guard let callback = callback else { return }
    
    guard let args = self.codec.decode(message) as? [Any?] else {
      callback(nil)
      return
    }
    
    let result = self.handler(args)
    callback(self.codec.encode(result))
  }
  
  func setMessageHandlerOnChannel(
    _ channel: String,
    binaryMessageHandler handler: FlutterBinaryMessageHandler? = nil
  ) -> FlutterBinaryMessengerConnection {
    self.count += 1
    return FlutterBinaryMessengerConnection(self.count)
  }
  
  func cleanUpConnection(_ connection: FlutterBinaryMessengerConnection) {
    // Method not implemented because this messenger is just for handling
  }
}
